src/meta/processors/admin/SnapShot.cpp

```diff
@@ -0,0 +1,197 @@
+#include "meta/processors/Common.h"
+#include "meta/ActiveHostsMan.h"
+#include "meta/MetaServiceUtils.h"
+#include "network/NetworkUtils.h"
+
+namespace nebula {
+namespace meta {
+cpp2::ErrorCode Snapshot::createSnapshot(const std::string& name) {
+    std::vector<GraphSpaceID> spaces;
+    kvstore::ResultCode ret = kvstore::ResultCode::SUCCEEDED;
+    if (!getAllSpaces(spaces, ret)) {
+        LOG(ERROR) << "Can't access kvstore, ret = d"
+                   << static_cast<int32_t>(ret);
+        return cpp2::ErrorCode::E_STORE_FAILURE;
+    }
+    for (auto& space : spaces) {
+        auto status = client_->createSnapshot(space, name).get();
+        if (!status.ok()) {
+            return cpp2::ErrorCode::E_RPC_FAILURE;
+        }
+    }
+
+    auto code = kv_->createCheckpoint(kDefaultSpaceId, name);
+    if (code != nebula::kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "Checkpoint create error on meta engine";
+        return cpp2::ErrorCode::E_STORE_FAILURE;
+    }
+
+    return cpp2::ErrorCode::SUCCEEDED;
+}
+
+cpp2::ErrorCode Snapshot::dropSnapshot(const std::string& name,
+                                       const std::vector<HostAddr> hosts) {
+    // The drop checkpoint will be skip if original host has been lost.
+    auto activeHosts = ActiveHostsMan::getActiveHosts(kv_);
+    std::vector<HostAddr> realHosts;
+    for (auto& host : hosts) {
+        if (std::find(activeHosts.begin(), activeHosts.end(), host) != activeHosts.end()) {
+            realHosts.emplace_back(host);
+        }
+    }
+
+    std::vector<GraphSpaceID> spaces;
+    kvstore::ResultCode ret = kvstore::ResultCode::SUCCEEDED;
+    if (!getAllSpaces(spaces, ret)) {
+        LOG(ERROR) << "Can't access kvstore, ret = d"
+                   << static_cast<int32_t>(ret);
+        return cpp2::ErrorCode::E_STORE_FAILURE;
+    }
+    for (auto& space : spaces) {
+        auto status = client_->dropSnapshot(space, name, realHosts).get();
+
+        if (!status.ok()) {
+            return cpp2::ErrorCode::E_RPC_FAILURE;
+        }
+    }
+    return cpp2::ErrorCode::SUCCEEDED;
+}
+
+bool Snapshot::getAllSpaces(std::vector<GraphSpaceID>& spaces, kvstore::ResultCode& retCode) {
+    // Get all spaces
+    folly::SharedMutex::ReadHolder rHolder(LockUtils::spaceLock());
+    auto prefix = MetaServiceUtils::spacePrefix();
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = kv_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        retCode = ret;
+        return false;
+    }
+    while (iter->valid()) {
+        auto spaceId = MetaServiceUtils::spaceId(iter->key());
+        spaces.push_back(spaceId);
+        iter->next();
+    }
+    return true;
+}
+
+cpp2::ErrorCode Snapshot::blockingWrites(storage::cpp2::EngineSignType sign) {
+    auto blk = (storage::cpp2::EngineSignType::BLOCK_OFF == sign) ? false : true;
+    auto code = kv_->setPartBlocking(kDefaultSpaceId, kDefaultPartId, blk);
```

Comment:
Should we block meta writing?

Reply:
I think best to block meta writing together. Because during the create snapshot,  we also need to blocking balance、set config , etc.  WDYT?

Reply:
If you block all writes for metad,  it means all storaged heartbeat will be expired.

Reply:
You're right. I missed the heartbeat. Let me solve it.

Reply:
Resoved the meta blocking problem. Now , the meta-server can be create checkpoint directly without blocking writes.